### PR TITLE
Ethernet driver for TI Hercules RM46

### PR DESCRIPTION
**Pull request by ChandranshuMalhotra**
Ethernet driver for TI Hercules RM46, to enable the board to communicate via TCP/IPv4 stack.

### CODE REVIEW

File: source/portable/NetworkInterface/RM46L852/NetworkInterface.c [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+ */
+
+BaseType_t xNetworkInterfaceInitialise( void )
+{
+    BaseType_t xFirstCall = pdTRUE;
+    BaseType_t xTaskCreated;
+
+    if (EMACHWInit(emacAddress) == EMAC_ERR_OK)
+    {
+        if( ( xFirstCall == pdTRUE ) || ( receiveTaskHandle == NULL ) )
```
Comment (Skptak) [inline source/portable/NetworkInterface/RM46L852/NetworkInterface.c:78]: You create and set `xFirstCall` inside of this function. This check will always return true.

File: source/portable/NetworkInterface/RM46L852/NetworkInterface.c [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+      xpbuf.len = pxNetworkBuffer->xDataLength;
+      xpbuf.tot_len = pxNetworkBuffer->xDataLength;
+      xpbuf.next = NULL;
+      if(xpbuf.tot_len < MIN_PKT_LEN)
+      {
+          xpbuf.tot_len = MIN_PKT_LEN;
+          xpbuf.len = MIN_PKT_LEN;
+      }
+
+      taskENTER_CRITICAL();
```
Comment (Skptak) [inline source/portable/NetworkInterface/RM46L852/NetworkInterface.c:126]: At time of writing only 4 Network Interfaces in [FreeRTOS-Plus-TCP/source/portable/NetworkInterface](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/tree/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface) have calls to enter or exit critical sections
The [STM32Hxx/NetworkInterface.c](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/STM32Hxx/NetworkInterface.c#L995) interface calls it exclusively to clear option bits, which is a few assembly instructions.
The other two interfaces, the [Zynq/x_emacspif_hw.c:HandleTxErrors()](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/Zynqx_emacpsif_hw.c#L235) and the [xilinx_ultrascale/x_emacpsif_hw.c:HandleTxErrors()]  **which actually both have the calls commented out.** Where it appears to be there for people to uncomment if attempting to de-bug an issue with a transmit error. 
Lastly the [ThirdParty/MSP432/NetworkInterface.c](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/bb654636d2cccd1121aeef681f005dfe8e38ca1a/source/portable/NetworkInterface/ThirdParty/MSP432/NetworkInterface.c#L788) which does enter and exit critical sections for its transfers. However, it does this in the task code that executes the packet transfers, not directly in the interface input and output calls. Additionally, this interface has a disclaimer about its use being untested with the current version of the FreeRTOS-Plus-TCP Stack.

**Why does this interface require the use of critical sections to transmit data when no other supported interfaces do?** This approach seems extremely heavy handed as it requires the end user of this interface to essentially set the networking actions as the single most import activity on the system.

File: source/portable/NetworkInterface/RM46L852/NetworkInterface.c [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            {
+                vReleaseNetworkBufferAndDescriptor( pxBufferDescriptor );
+                iptraceETHERNET_RX_EVENT_LOST();
+                FreeRTOS_printf( ( "RX Event Lost\n" ) );
+            }
+        }
+        else
+        {
+            #if ( ( ipconfigHAS_DEBUG_PRINTF == 1 ) && defined( FreeRTOS_debug_printf ) )
+                const EthernetHeader_t * pxEthernetHeader;
```
Comment (Skptak) [inline source/portable/NetworkInterface/RM46L852/NetworkInterface.c:186]: It's against C coding conventions to declare variables in the middle of the function. These variables should be wrapped in an ifdef at the start of the function.

Comment (htibosch) [reply source/portable/NetworkInterface/RM46L852/NetworkInterface.c:186]: Hello @Skptak, as the variable ` pxEthernetHeader` is only used for logging, I prefer to see a local declaration.

Normally we ( in FreeRTOS+TCP ) adhere to the rule that local declarations are OK, unless they're mixed with code, They should be placed at the beginning of a compound block:

~~~
    if ()
    {
        unsigned rc;  // good
        counter++;
    }
    if ()
    {
        counter++;
        unsigned rc;  // bad: code followed by a declaration
    }
~~~

The advantage is that the code will be easier to read. The reader doesn't have don't have to scroll back to see the declarations. Another advantage is that the scope of the variable is limited when declared in a local block.

**Pull request closed without merging**